common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -511,73 +521,372 @@ private ImmutableSetMultimap<Class<? extends Annotation>, Element> toClassKeyedM
   }
 
   /**
-   * A package or type name.
+   * An {@link ElementName} for an annotated element.
    *
-   * <p>It's unfortunate that we have to track types and packages separately, but since there are
-   * two different methods to look them up in {@link Elements}, we end up with a lot of parallel
-   * logic. :(
+   * <p>Instead of saving elements, an {@code ElementName} is saved since
    *
-   * <p>Packages declared (and annotated) in {@code package-info.java} are tracked as deferred
-   * packages, type elements are tracked directly, and all other elements are tracked via their
-   * nearest enclosing type.
+   * <ol>
+   *   <li>There is no guarantee that any particular element will always be represented by the same
+   *       object. (Reference: {@link Element})
+   *   <li>Since an implementation may choose to have a single object implement multiple Element
+   *       sub-interfaces. (Reference: {@link Element})
+   *   <li>and possibly other issues (such as reconstruction of AST after each round, and direct
+   *       manipulations of AST by tools like lombok) may make the saved element reference
+   *       unreliable.
+   * </ol>
    */
-  private static final class ElementName {
-    private enum Kind {
-      PACKAGE_NAME,
-      TYPE_NAME,
+  private abstract static class ElementName {
+
+    private ElementName(Element element) {
+      Objects.requireNonNull(element);
+      if (!isAcceptableElementKind(element)) {
```

Comment:
I have removed the check; however, not the `isAcceptableElementKind()` method because I believe it makes the code more symmetric and readable. Please let me know what you think.